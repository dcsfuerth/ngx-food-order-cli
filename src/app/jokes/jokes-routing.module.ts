import { NgModule } from '@angular/core';
import { RouterModule, Routes } from '@angular/router';
import { JokesPageComponent } from './jokes-page/jokes-page.component';

const routes: Routes = [{ path: '', component: JokesPageComponent }];

@NgModule({
  imports: [RouterModule.forChild(routes)],
  exports: [RouterModule],
})
export class JokesRoutingModule {}
