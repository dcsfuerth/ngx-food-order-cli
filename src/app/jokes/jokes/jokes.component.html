<div class="row">
  <div class="col">
    <h2>Randon Chuck Norris Jokes</h2>
    <ul class="list-group">
      <li *ngFor="let joke of jokes;" class="list-group-item">
        <p [innerHTML]="joke.joke"></p>
        <small>{{ joke.categories }}</small>
      </li>
    </ul>

    <dcs-loading-indicator [loading]="loading"></dcs-loading-indicator>
  </div>
</div>
