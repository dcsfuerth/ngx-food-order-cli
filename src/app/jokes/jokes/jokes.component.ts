import { ChangeDetectionStrategy, Component, Input } from '@angular/core';
import { Joke } from '../../reducers/jokes/joke.class';

@Component({
  selector: 'dcs-jokes',
  templateUrl: './jokes.component.html',
  styleUrls: ['./jokes.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class JokesComponent {
  @Input() jokes: Joke[];
  @Input() loading: boolean;
}
