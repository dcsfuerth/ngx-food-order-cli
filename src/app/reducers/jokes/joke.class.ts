import { ViewModel } from '@dcs/ngx-tools';

export interface IJoke {
  id: string;
  joke: string;
  categories: string[];
}

export class Joke extends ViewModel<IJoke> {
  protected static readonly defaults: IJoke = {
    id: '',
    joke: '',
    categories: ['nerdy'],
  };

  public id: string;
  public joke: string;
  public categories: string[];

  constructor(props: Partial<IJoke> = {}) {
    props = { ...Joke.defaults, ...props };
    super(props);
  }
}
