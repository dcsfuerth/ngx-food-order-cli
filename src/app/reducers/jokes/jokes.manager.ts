import { State } from '..';
import { schema } from 'normalizr';
import { IJoke, Joke } from './joke.class';
import {
  INormalizedCollectionState,
  generateNormalizedCollectionState,
  readOnlyCollectionManagerFactory,
} from '@dcs/ngx-tools';

export interface IJokesState extends INormalizedCollectionState {
  entities: { jokes: { [key: string]: IJoke } };
}

const initialState = generateNormalizedCollectionState<IJokesState>({ jokes: {} });

const jokeSchema = new schema.Entity(
  'jokes',
  {},
  {
    idAttribute(entity) {
      return String(entity.id);
    },
  }
);

const jokesSchema = new schema.Array(jokeSchema);

export const jokesStateSelector = (state: State): IJokesState => state.jokes;

export const jokesManager = readOnlyCollectionManagerFactory(
  'Jokes',
  initialState,
  jokesStateSelector,
  jokesSchema,
  Joke
);
