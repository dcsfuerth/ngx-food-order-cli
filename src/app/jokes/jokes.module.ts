import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { JokesPageComponent } from './jokes-page/jokes-page.component';
import { JokesRoutingModule } from './jokes-routing.module';
import { JokesComponent } from './jokes/jokes.component';
import { ComponentsModule } from '../components/components.module';

@NgModule({
  imports: [CommonModule, JokesRoutingModule, ComponentsModule],
  declarations: [JokesPageComponent, JokesComponent],
})
export class JokesModule {}
