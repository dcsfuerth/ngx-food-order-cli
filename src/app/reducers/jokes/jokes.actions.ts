import { ApiGetRequest } from '@dcs/ngx-tools';
import { IJoke } from './joke.class';
import { jokesManager } from './jokes.manager';

export interface IJokesResponse {
  type: string;
  value: IJoke[];
}

export class FetchJokes extends ApiGetRequest {
  constructor(count: number = 5) {
    super(
      `http://api.icndb.com/jokes/random/${count}`,
      jokesManager.actions.fetch.base,
      jokesManager.schema,
      undefined,
      // the real data is embedded, unpack before giving it
      // to the normalized tool chain
      (data: IJokesResponse) => data.value
    );
  }
}
