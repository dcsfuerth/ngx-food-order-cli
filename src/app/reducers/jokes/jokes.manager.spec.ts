import { Joke } from './joke.class';
import { jokesManager } from './jokes.manager';

describe('Jokes Manager', () => {
  describe('initialState', () => {
    it('has the correct format', () => {
      expect(jokesManager.initialState).toMatchSnapshot();
    });
  });

  describe('reducer', () => {
    it('returns the initialState', () => {
      const newState = jokesManager.reducer(undefined, { type: 'any' });
      expect(newState).toMatchSnapshot();
      expect(newState).toEqual(jokesManager.initialState);
    });

    it('sets loading to true on start action', () => {
      const newState = jokesManager.reducer(undefined, { type: jokesManager.actions.fetch.start });
      expect(newState).toMatchSnapshot();
    });
  });

  describe('selectors', () => {
    const state = {
      jokes: {
        ...jokesManager.initialState,
        loaded: true,
        result: ['42'],
        entities: { jokes: { 42: { id: '42', joke: 'Knock, knock!', categories: ['bad jokes'] } } },
      },
    };

    describe('collection', () => {
      it('returns a collection of Joke instances', () => {
        const collection = jokesManager.selectors.collection(state);
        expect(collection[0]).toEqual(expect.any(Joke));
        expect(collection).toMatchSnapshot();
      });
    });

    describe('loaded', () => {
      it('returns true', () => {
        expect(jokesManager.selectors.loaded(state)).toEqual(true);
      });
    });

    describe('loading', () => {
      it('returns false', () => {
        expect(jokesManager.selectors.loading(state)).toEqual(false);
      });
    });
  });
});
