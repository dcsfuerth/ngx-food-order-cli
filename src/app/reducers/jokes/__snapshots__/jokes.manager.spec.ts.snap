// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Jokes Manager initialState has the correct format 1`] = `
Object {
  "entities": Object {
    "jokes": Object {},
  },
  "error": null,
  "lastState": null,
  "loaded": false,
  "loading": false,
  "result": Array [],
  "updatedAt": null,
  "updating": false,
}
`;

exports[`Jokes Manager reducer returns the initialState 1`] = `
Object {
  "entities": Object {
    "jokes": Object {},
  },
  "error": null,
  "lastState": null,
  "loaded": false,
  "loading": false,
  "result": Array [],
  "updatedAt": null,
  "updating": false,
}
`;

exports[`Jokes Manager reducer sets loading to true on start action 1`] = `
Object {
  "entities": Object {
    "jokes": Object {},
  },
  "error": null,
  "lastState": null,
  "loaded": false,
  "loading": true,
  "result": Array [],
  "updatedAt": null,
  "updating": false,
}
`;

exports[`Jokes Manager selectors collection returns a collection of Joke instances 1`] = `
Array [
  "{\\"id\\":\\"42\\",\\"joke\\":\\"Knock, knock!\\",\\"categories\\":[\\"bad jokes\\"]}",
]
`;
